### PR TITLE
senku: sigmoid-l1 slice attention vs softmax (architecture A/B)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Transolver (the architecture family our 4L/512d model uses) computes **slice attention**: each input point softly assigns probability over `K` slices via a softmax over a slice-logit projection. Each slice computes a context vector, and the per-point representation is read out as a softmax-weighted combination.

The softmax slicing creates a **winner-take-all dynamic**: most of a point's probability mass concentrates on 1–2 slices, and gradients to "non-selected" slices are tiny. This may be problematic for CFD where many points lie at slice boundaries (e.g. wake/free-stream interface, separated-flow regions) and would benefit from richer mixed-slice membership.

**Hypothesis**: replace the softmax slice assignment with a **sigmoid gate** (independent per-slice soft membership in [0,1] for each point), normalized to sum to 1 per point only at readout. This:

1. Allows multi-slice membership (e.g. a wake-edge point can belong to both "wake" and "free-stream" slices simultaneously).
2. Keeps gradient flow to all slices regardless of which is dominant — better signal during training.
3. Increases effective slice diversity (no winner-take-all collapse).

This is a small architectural change with strong literature precedent (sigmoid attention has been shown competitive with softmax attention in Tay 2023, "FLASH"; gated mixture-of-experts uses similar mechanics).

**Risk**: sigmoid attention with no-normalization baseline can blow up in scale. The fix is to L1-normalize the gates per-point at readout (so the slice mixture stays a convex-like combination): `gates_normalized = gates / (gates.sum(dim=-1, keepdim=True) + eps)`. This preserves the "convex combination" reading at output while allowing the slice-logits to express richer multi-membership during attention.

## Instructions

**Code change required.** Add a flag `--slice-attention <softmax|sigmoid-l1>` (default `softmax` = current behavior) to `target/train.py`. When set to `sigmoid-l1`:

1. Locate the slice-attention block. Search `train.py` for the slice-logit softmax — typically something like `slice_weights = F.softmax(slice_logits, dim=-1)` or similar in the forward of the Transolver attention layer.
2. Replace with:

```python
if self.config.slice_attention == "softmax":
    slice_weights = F.softmax(slice_logits, dim=-1)
elif self.config.slice_attention == "sigmoid-l1":
    gates = torch.sigmoid(slice_logits)
    slice_weights = gates / (gates.sum(dim=-1, keepdim=True) + 1e-6)
else:
    raise ValueError(f"unknown slice_attention: {self.config.slice_attention}")
```

3. The rest of the slice attention block (slice-context computation, per-point readout) is **unchanged** — only the gating function changes. This means everything that took `slice_weights` as an input continues to work.

4. **Verify** by logging a per-batch diagnostic at training start: the **mean L1 norm of `gates`** before normalization (call it `train/sigmoid_gate_l1_mean`). With softmax this would always be 1.0; with sigmoid it should be roughly K/2 = 64 at init (random logits → sigmoid ≈ 0.5). If it stays ≈ K/2 throughout training, the model is using all slices uniformly (good — signal of multi-membership). If it collapses toward 1.0 (winner-take-all), the sigmoid is recovering softmax behavior (signal of convergence to softmax-like solution).

**Run a 2-arm DDP-4 sweep on a single 4-GPU pod, sequential arms:**

| Arm | `--slice-attention` | rationale |
|---|---|---|
| A | `softmax` (control) | in-group reference |
| B | `sigmoid-l1` | multi-slice gated membership |

(Two arms is enough for a head-to-head. If B clearly wins, follow-ups can sweep alternatives like `sigmoid-l2`, temperature scaling, etc.)

**Common config (Lion SOTA recipe, 128 slices):**

```bash
cd target/
torchrun --standalone --nproc_per_node=4 train.py \
  --agent senku \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 1e-4 \
  --clip-grad-norm 0.5 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --seed 42 \
  --wandb-group yi-r31-senku-slice-attention
```

Per-arm: `--slice-attention softmax` / `sigmoid-l1`.

No kill thresholds. Each arm gets the full 6h.

**Optional Arm C** (if both A/B finish under budget): `--slice-attention sigmoid-l1 --model-slices 64` to test whether multi-membership lets us use fewer slices for the same expressivity. Skip if time-bounded.

## What to report

Per-arm:
1. Best `val_primary/abupt_axis_mean_rel_l2_pct`, with corresponding test row.
2. Per-channel `surface_pressure`, `volume_pressure`, `wall_shear`, `wall_shear_y`, `wall_shear_z`.
3. **Critical diagnostic**: `train/sigmoid_gate_l1_mean` time series across training. Did the gates stay at ~K/2 (multi-membership) or collapse toward 1 (winner-take-all)?
4. **Slice usage entropy**: histogram of slice-assignment entropy across a held-out validation batch. Higher entropy = more multi-slice membership.
5. W&B run IDs.

## Decision rule (advisor side)

- Any arm beats yi merge bar **9.039%** → merge candidate.
- Sigmoid arm wins → multi-slice membership pays off; assign next-round refinements (alternative normalizations, learnable slice temperatures).
- Sigmoid arm draws and gates collapse to softmax behavior → null result, mechanism not exercised, close.
- Sigmoid arm clearly loses → winner-take-all is *desired* in this regime; close and document.

## Baseline

- **Active merge bar:** `val_primary/abupt_axis_mean_rel_l2_pct < 9.039%` (PR #309 thorfinn).
- **Aspirational once PR #490 STRING-sep PE lands:** `< 7.546%`.

## Notes

- Code change is small (~15 lines): one flag + one branch on the gating function + one diagnostic log.
- This is an **architecture experiment**, not a hyperparameter one. The advisor (per CLAUDE.md plateau protocol) is encouraging bold architectural moves.
- DDP-4, bs=4, 6h per arm → ~10 epochs. 2 arms × 6h = 12h. Sequential.
- Extension space (queued for follow-up depending on result): scaled sigmoid (`sigmoid(s · logits)` with learnable s), entmax-1.5 (interpolates between softmax and sparsemax), per-head separate gating function.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: approved. Sigmoid-L1 vs softmax slice attention is exactly the kind of architectural intervention to take while we're past the easy hyperparameter wins. The winner-take-all critique of softmax slicing is well-motivated for CFD where slice-boundary regions (wake/free-stream interfaces, separation lines) are physically the hardest points to predict.

**Updated baseline:** the active merge bar is now **val_abupt 7.1792%** (PR #489 thorfinn, vol-points curriculum, run `r5rw40rn`), not 9.039%. The 9.039% in your PR body is stale (PR #309). Update your decision rule accordingly.

**Locating the slice softmax:** the Transolver slice block is in the model file (search for `F.softmax` near the slice-logits projection in `SliceAttention` or whatever the class is named). Likely 1–2 occurrences total. Be careful: there may be a per-head softmax separate from the slice-logits softmax — make sure you only swap the slice-assignment one, not the standard inner-attention QK-softmax.

**Diagnostic logging is the most valuable part of this experiment.** `train/sigmoid_gate_l1_mean` time-series tells us whether sigmoid is *exercising* the multi-membership mechanism or collapsing back to softmax behavior. Add a second diagnostic: per-batch slice usage histogram entropy at training mid-point and end. If entropy stays high → multi-membership is real; if it collapses → null result, and we know mechanism wasn't exercised.

**Risk note:** sigmoid + L1-norm is mathematically a softmax with a learnable temperature implicit in the magnitude scale. The gradient signal differs but the expressive class is similar. If Arm B essentially ties Arm A on val_abupt, that's the expected outcome and not a failure — it just means the inductive bias of softmax wasn't a constraint here.

**Skip Arm C (sigmoid + 64 slices)** unless A and B both finish well under budget. Two clean arms is more informative than three half-trained ones.

Ready when you are. Mark ready-for-review with both diagnostic plots attached as W&B image artifacts or screenshots.

**Comment by morganmcg1** [comment]
ADVISOR CORRECTION: my previous comment cited val_abupt 7.1792% from PR #489 as the active merge bar. That was an error — **PR #489 (thorfinn, vol-points curriculum) was merged into the `tay` branch, not `yi`**. The active yi merge bar is unchanged: **val_abupt 9.039%** (PR #309 thorfinn grad-clip=0.5).

7.1792% (PR #489) and 7.546% (PR #311) are aspirational tay-branch numbers, not currently reproducible on the yi codebase. Beat 9.039% to merge. Apologies for the confusion.

**Comment by morganmcg1** [comment]
STUDENT senku:

## Results

**Verdict: Sigmoid-L1 multi-membership gating clearly LOSES to softmax winner-take-all on the 4L/512d/128-slice Lion stack.**

Per the PR decision rule:
> Sigmoid arm clearly loses → winner-take-all is *desired* in this regime; close and document.

### Headline metrics (both arms ran 3 epochs, best @ epoch 3)

| Arm | val_primary | test_primary | Δ vs Arm A (val / test) |
|---|---|---|---|
| **A** (softmax, control) | **11.278%** | **12.403%** | reference |
| **B** (sigmoid-l1) | 13.366% | 14.432% | **+2.087pp / +2.029pp (worse)** |

Neither arm beat the active 9.039% yi merge bar (PR #309) — these were 4-GPU 6h-budget runs (~3 epochs), well short of the full-budget Lion+warmup recipe used to set that bar. The head-to-head is the conclusion that matters here, not the absolute number.

### Per-channel breakdown (full_val_primary / test_primary, %)

| Channel | A val | B val | Δ val | A test | B test | Δ test |
|---|---:|---:|---:|---:|---:|---:|
| surface_pressure | 7.228 | 9.037 | +1.81 | 7.026 | 8.901 | +1.88 |
| wall_shear_x | 10.659 | 12.504 | +1.85 | 10.650 | 12.497 | +1.85 |
| wall_shear_y | 15.698 | 17.625 | +1.93 | 15.548 | 17.361 | +1.81 |
| wall_shear_z | 16.065 | 19.764 | **+3.70** | 15.411 | 19.043 | **+3.63** |
| volume_pressure | 6.741 | 7.898 | +1.16 | 13.382 | 14.357 | +0.98 |
| **abupt_axis_mean** | **11.278** | **13.366** | **+2.09** | **12.403** | **14.432** | **+2.03** |

Sigmoid-L1 regresses on **every channel**; the worst hit is wall_shear_z (+3.7pp), the channel with the noisiest/most-anisotropic signal. Volume pressure (smoothest) takes the smallest hit — opposite of the hypothesis that multi-membership would help heterogeneous channels most.

### Mechanistic diagnostics — Arm B (sigmoid-l1) gate trajectory

The sigmoid gates were exercised, did not collapse to softmax winner-take-all, but did concentrate substantially:

| step | gate_l1_mean | gate_top1_mean | active>0.5 | entropy_norm | train_loss |
|---:|---:|---:|---:|---:|---:|
| 2 (init) | 63.93 | 0.59 | 62.5 | 0.9995 | 2.087 |
| 1114 | 61.79 | 0.86 | 63.0 | 0.9625 | 1.058 |
| 2642 | 57.45 | 0.96 | 58.7 | 0.8864 | 0.340 |
| 5478 | 43.28 | 0.99 | 43.6 | 0.7655 | 0.111 |
| 8200 | 40.07 | 0.98 | 40.4 | 0.7288 | 0.065 |
| 10778 (end) | 37.98 | 0.98 | 38.2 | 0.7060 | 0.036 |

Reference values: K=128, init `l1_mean = K/2 = 64`, init `entropy_norm ≈ 1.0`. Winner-take-all (softmax-like) would push `l1_mean → 1`, `top1 → 1`, `active>0.5 → 1`, `entropy_norm → 0`.

**Interpretation:** The gates settled into "wide multi-membership over ≈30% of the slice budget" — by training end, ≈38 of 128 slices have gate>0.5 per point, with one nearly-saturated leader (top1≈0.98) plus ~37 mid-active companions. This is exactly the multi-slice membership the hypothesis predicted, just substantially narrower than init: the model wanted *some* concentration, used the freedom of the sigmoid to drop ~70% of the slices entirely, but kept the remaining 30% live. Entropy stayed at 0.71 (not 0 collapse), confirming the mechanism was active throughout.

So this is **not a "the mechanism never engaged, null result" outcome** — sigmoid-L1 did learn a meaningfully different slice-membership pattern than softmax would. That pattern was just worse for the task.

### Why softmax wins here (best read of the evidence)

1. **128 slices is plenty of capacity.** With softmax, each point picks 1–2 slices and the *slices themselves* specialize across the whole dataset. The model effectively has 128 specialist functions and a router. Sigmoid-L1 mixes ~38 of them per point, which dilutes specialist gradients (38× more participation per slice → 38× lower selection sharpness for any one specialist).
2. **The L1 renormalization punishes multi-membership at readout.** A point keeping 38 active gates ends up with each slice contributing ~1/38 of the readout vector — the magnitude of any single specialist's contribution is small, so individual-slice expressiveness is reduced.
3. **CFD slices want sharp regional specialization.** Wake/shear/free-stream regions have crisp boundaries (the body surface is geometric, not statistical). The "many points lie at slice boundaries" intuition appears to be wrong for DrivAerML at this scale — geometry-driven slicing converges quickly under softmax. wall_shear_z (most localized signal) being hit hardest supports this.

### Run details

**Arm A (softmax, control)** — W&B run `zd1cmbvj`
- Group: `yi-r31-senku-slice-attention`
- Wall time: ~4h54m (3 epochs at 17m/epoch incl. validation)
- Final config: `--slice-attention softmax`

**Arm B (sigmoid-l1)** — W&B run `6cberadx`
- Group: `yi-r31-senku-slice-attention`
- Wall time: ~5h01m (3 epochs)
- Final config: `--slice-attention sigmoid-l1`

Exact `train.py` commands used (identical except for `--slice-attention`):

```bash
cd target/
torchrun --standalone --nproc_per_node=4 train.py \
  --agent senku \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 1e-4 \
  --clip-grad-norm 0.5 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --seed 42 \
  --slice-attention <softmax | sigmoid-l1> \
  --wandb-group yi-r31-senku-slice-attention \
  --wandb-name senku/r31-slice-attn-<A-softmax | B-sigmoid-l1>
```

**Peak memory:** ~70.7 GB / GPU per `nvidia-smi` during training (RTX PRO 6000 Blackwell, 4-GPU DDP, model 12.70M params); identical between arms (the sigmoid+L1 fp32 ops add a few hundred MB at most, well within the cushion).

### What happened

The hypothesis was that multi-slice membership would help points near slice boundaries (e.g. wake-edge) by maintaining gradient flow to several relevant slices simultaneously. The mechanism *did* engage — the sigmoid gates did not collapse to softmax-like winner-take-all and ended training with ~38 active slices per point on average — but the resulting representation regressed on every channel by 1.0–3.7pp. The largest regression was on `wall_shear_z`, which is the channel with the most localized/anisotropic surface signal, exactly opposite of what the hypothesis predicted (multi-membership was supposed to help heterogeneous channels most).

The clean read is that **sharp specialization is a feature, not a bug** for the slice-attention layer in this 4L/512d/K=128 regime. Softmax forces 128 slices to specialize tightly; that specialization is what carries the representation. Sigmoid-L1's wider per-point coverage dilutes each specialist's contribution at readout (1/L1 normalization shrinks each gate when many are active) and slows specialization formation during training.

### Suggested follow-ups

Per the decision rule this should close, but recording follow-ups in case the advisor wants to push the design space:

1. **Scaled sigmoid `sigmoid(s · logits)` with learnable `s`.** Lets the network learn its own sharpness — equivalent to softmax temperature but without forced sum-to-1. Might recover the specialization sharpness while keeping the gradient-flow benefit. (Mentioned in PR body as extension.)
2. **Sparser gating (top-K sigmoid or entmax-1.5).** Forces only K active gates per point. Picks the multi-membership granularity by hand instead of learning it. K=2 or K=4 likely closer to the softmax sweet spot.
3. **Sigmoid-L1 with smaller K (e.g. K=64 or K=32).** With multi-membership, fewer slices may suffice. The PR's optional Arm C — skipped here for time. Worth doing only if a pre-screen ablation suggests sigmoid-L1 was capacity-constrained, which the diagnostics here don't support (final L1≈38 < init K/2=64 means the model was actively *reducing* slice count it considered, not capacity-starved).
4. **Per-head separate gating function.** Each head picks its own softmax/sigmoid. The 8 heads might want different sharpness — but this is more flag plumbing for marginal expected upside, lower priority.

For round 31 onward, recommend closing this experiment as a clear negative (winner-take-all is desired in this regime) and pivoting compute to other architectural moves with more upside per hour.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #521 because ADVISOR: Closing — null result. Arm B sigmoid-l1 slice attention val=13.366% vs Arm A softmax control val=11.278% (delta +2.09pp). Per the PR's own decision rule: 'Sigmoid arm clearly loses → winner-take-all is desired in this regime; close and document.' This confirms the Transolver softmax slice assignment is the correct inductive bias for DrivAerML at our config — multi-slice membership dilutes the slice-token routing signal. Logging this as a closed dead end so we don't revisit. Thanks for the clean two-arm A/B..
